File: server/app/models/report.py
from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Integer, Text, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from server.app.db.base import Base

if TYPE_CHECKING:
    from server.app.models.notification import Notification


class Report(Base):
    __tablename__ = "reports"
    __table_args__ = (UniqueConstraint("report_type", "period_start", "period_end", name="uq_reports_type_period"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    report_type: Mapped[str] = mapped_column(Text, nullable=False)
    period_start: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    period_end: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    status: Mapped[str] = mapped_column(Text, nullable=False, server_default="generated")
    subject: Mapped[str] = mapped_column(Text, nullable=False)
    summary: Mapped[str | None] = mapped_column(Text)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    hotspot_count: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())

    notifications: Mapped[list[Notification]] = relationship(back_populates="report")

File: internal/model/entity/report.go
package entity

import "time"

type Report struct {
	ID           int64      `gorm:"column:id;primaryKey"`
	UserID       int64      `gorm:"column:user_id"`
	ReportType   string     `gorm:"column:report_type"`
	PeriodStart  time.Time  `gorm:"column:period_start;type:date"`
	PeriodEnd    time.Time  `gorm:"column:period_end;type:date"`
	Subject      string     `gorm:"column:subject"`
	Summary      string     `gorm:"column:summary"`
	Content      string     `gorm:"column:content"`
	HotspotCount int        `gorm:"column:hotspot_count"`
	Status       string     `gorm:"column:status"`
	SentAt       *time.Time `gorm:"column:sent_at"`
	CreatedAt    time.Time  `gorm:"column:created_at"`
	UpdatedAt    time.Time  `gorm:"column:updated_at"`
}

func (Report) TableName() string { return "reports" }

type ReportExport struct {
	ID           int64      `gorm:"column:id;primaryKey"`
	ReportID     int64      `gorm:"column:report_id"`
	ExportKind   string     `gorm:"column:export_kind"`
	TargetPath   string     `gorm:"column:target_path"`
	Status       string     `gorm:"column:status"`
	ErrorMessage string     `gorm:"column:error_message"`
	PublishedAt  *time.Time `gorm:"column:published_at"`
	CreatedAt    time.Time  `gorm:"column:created_at"`
	UpdatedAt    time.Time  `gorm:"column:updated_at"`
}

func (ReportExport) TableName() string { return "report_exports" }
